tools/fileinfo_emitter/src/main.rs

```diff
@@ -0,0 +1,298 @@
+struct PosInfo {
+    #[serde(serialize_with = "id_serialize_passthrough")]
+    pub name: Id,
+    pub filename: String,
+    pub linenum: usize,
+    pub varname: String,
+}
+
+struct ComponentPosIds {
+    pub component_pos_id: u32,
+    pub cells: HashMap<Id, u32>,
+    pub groups: HashMap<Id, u32>,
+}
+
+fn gen_component_info(
+    ctx: &ir::Context,
+    comp: &ir::Component,
+    component_info: &mut HashMap<Id, ComponentPosIds>,
+) -> CalyxResult<()> {
+    // get pos for component
+    let component_set_attr = comp
+        .attributes
+        .get_set(SetAttribute::Set(SetAttr::Pos))
+        .unwrap();
+    let component_pos = component_set_attr.iter().next().unwrap();
+    let mut component_pos_id = ComponentPosIds {
+        component_pos_id: *component_pos,
+        cells: HashMap::new(),
+        groups: HashMap::new(),
+    };
+
+    // get pos for groups
+    for group in comp.groups.iter() {
+        let group_ref = group.borrow();
+        let group_set_attr = group_ref
+            .attributes
+            .get_set(SetAttribute::Set(SetAttr::Pos))
+            .unwrap();
+        let group_pos = group_set_attr.iter().next().unwrap();
+        component_pos_id
+            .groups
+            .insert(group.borrow().name(), *group_pos);
+    }
+
+    // get pos for cell
+    for cell in comp.cells.iter() {
+        let cell_ref = cell.borrow();
+        match cell_ref.attributes.get_set(SetAttribute::Set(SetAttr::Pos)) {
+            None => {
+                dbg!("Ignoring cell without pos: {}", cell_ref.name());
+                continue;
+            }
+            Some(cell_set_attr) => {
+                let cell_pos = cell_set_attr.iter().next().unwrap();
+                component_pos_id.cells.insert(cell_ref.name(), *cell_pos);
+                if let ir::CellType::Component { name } = cell_ref.prototype {
+                    let component = ctx
+                        .components
+                        .iter()
+                        .find(|comp| comp.name == name)
+                        .unwrap();
+                    gen_component_info(ctx, component, component_info)?;
+                }
+            }
+        }
+    }
+    component_info.insert(comp.name, component_pos_id);
+    Ok(())
+}
+
+fn obtain_pos_info(
+    name: &Id,
+    pos_id: &u32,
+    source_info_table: &SourceInfoTable,
+    file_lines_map: &HashMap<String, Vec<String>>,
+) -> CalyxResult<PosInfo> {
+    let SourceLocation { file, line } =
+        source_info_table.lookup_position(PositionId::from(*pos_id));
+    let filename = source_info_table
+        .lookup_file_path(*file)
+        .as_path()
+        .to_str()
+        .unwrap();
+    Ok(PosInfo {
+        name: *name,
+        filename: filename.to_string(),
+        linenum: line.as_usize(),
+        varname: get_var_name(
+            &filename.to_string(),
+            line.as_usize(),
+            file_lines_map,
+            name,
+        ),
+    })
+}
+
+fn get_var_name(
```

Comment:
what does this function do? is the information not available from the calyx parser?

Reply:
I updated the function name and added some documentation, but this function attempts to fetch the ADL-level name of the component/cell/group (via scanning the line for a `=`).